### DISCUSSION TITLE
Crop or Pad Help

### DISCUSSION THREAD

**Discussion by ik362** [TorchIO-project/torchio#576] [Q&A]
Hi there!

I am using torchio+pytorch-lightning to do some image classification. The tricky part is that I use 4D fMRI data; so far this has not been a problem except that not all my data is the same shape.

I wanted to ask then: is there a way that I can use CropOrPad on 4D data (looking at the docs for CropOrPad looks to only work in 3D)?

If CropOrPad cannot be applied to 4D data is there another workaround?

Cheers!

**Comment by fepegar** [comment]
Hi, @ik362.

I'm not sure what the problem is. `CropOrPad` does work for 4D data.

```python
import torch
import torchio as tio

fmri = tio.ScalarImage(tensor=torch.rand(35, 128, 128, 92))
transform = tio.CropOrPad((88, 88, 88))
transformed = transform(fmri)
```

Can you tell us a bit more about your data and your application?

**Reply by ik362** [reply]
Hi @fepegar,

Thanks for your reply! After a little more investigation I think I have a better idea of the problem.

For some context here is my preprocess code:
```python
def get_preprocessing_transform(self):
        preprocess = tio.Compose([
            tio.RescaleIntensity((-1, 1)),
            tio.CropOrPad((64, 64, 40)),
        ])
        return preprocess
```

And here is the pl.LigtningModule code:
```python
def prepare_batch(self, batch):
        return batch['image'][tio.DATA], batch['diagnosis']

def forward(self, x):
    encoded = self.encoder(x)
    decoded = self.decoder(encoded)
    return decoded
    
def training_step(self, batch, batch_idx):
    x, _ = self.prepare_batch(batch)
    y_hat = self.forward(x)
    train_loss = self.criterion(y_hat, x)
    self.log('train_loss', train_loss, prog_bar = True)
    return train_loss
```

When I look at the shape of x and y_hat I get:
```python
x shape: torch.Size([1, 64, 64, 64, 40])
y_hat shape: torch.Size([1, 64, 6, 6, 5])
```
So it seems that CropOrPad is padding the last 3 dimensions(?)

I think there are two questions:
1. Do I have to reshape images (e.g. `x = x.view(w, d, h, t, b)`) every time?
2. I didnt make my original question very clear (sorry) can I apply CropOrPad in the time domain? My scans have a different number of time points and because of my architecture I need them to all have the same time points

Thanks!

**Reply by fepegar** [reply]
Thanks for the context. I might need some more.

> So it seems that CropOrPad is padding the last 3 dimensions(?)

It's hard to know without knowing the shape of the input. Do you have 64 timepoints? 

> Do I have to reshape images (e.g. x = x.view(w, d, h, t, b)) every time?

Hmm I don't think you need this. The input and output of the transform are meant to be (channels, spatial1, spatial2, spatial3). In your case, the channels axis can be used for time. The batch dimension is added by the data loader you use.

> I didnt make my original question very clear (sorry) can I apply CropOrPad in the time domain? My scans have a different number of time points and because of my architecture I need them to all have the same time points

Oh, I see. Maybe you can reshape your images before and after `CropOrPad`, using a `Lambda` transform.

```python
In [1]: import torchio as tio

In [2]: !curl -o image.jpg https://thispersondoesnotexist.com/image
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  440k  100  440k    0     0  1002k      0 --:--:-- --:--:-- --:--:-- 1004k

In [3]: image = tio.ScalarImage('image.jpg')

In [4]: image.plot()
```

![image](https://user-images.githubusercontent.com/12688084/123553960-c7e8cb80-d775-11eb-89e9-71f4a790c49c.jpg)

```python
In [5]: image.set_data(image.data[0:1])  # keep one channel only

In [6]: image.shape
Out[6]: (1, 1024, 1024, 1)

In [7]: image.plot()
```

![tmpvfvghyb_](https://user-images.githubusercontent.com/12688084/123553998-f9fa2d80-d775-11eb-9709-4939a1feb383.png)

```python
In [8]: reshape1 = tio.Lambda(lambda x: x.permute(1, 2, 3, 0))

In [9]: reshaped = reshape1(image)

In [18]: reshaped.shape
Out[18]: (1024, 1024, 1, 1)

In [19]: pad = tio.CropOrPad((1024, 1, 3))

In [20]: padded = pad(reshaped)

In [21]: padded.shape
Out[21]: (1024, 1024, 1, 3)

In [21]: reshape2 = tio.Lambda(lambda x: x.permute(3, 0, 1, 2))

In [22]: green = reshape2(padded)

In [23]: green.shape
Out[23]: (3, 1024, 1024, 1)

In [24]: green.plot()
```

![tmpk_kl6blr](https://user-images.githubusercontent.com/12688084/123554262-05018d80-d777-11eb-91f8-2edbaf66e40d.png)

Does that make sense? It's a bit hacky, but it should work.

**Reply by ik362** [reply]
Hi @fepegar,

Thank you so much for your help! Based on your example I came up with this function:

```python
def trim_timepoints(image, frames):
    reshape1 = tio.Lambda(lambda x: x.permute(1, 2, 3, 0))
    reshape2 = tio.Lambda(lambda x: x.permute(3, 0, 1, 2))
    img = tio.ScalarImage(image)
    w, d, h, t = img[tio.DATA].size()
    crop = tio.CropOrPad((h, frames, w))
    reshaped = reshape1(img)
    cropped = crop(reshaped)
    new = reshape2(cropped)
    return new
```
Seemed to work on some dummy data during my testing; does this seem okay to you? 

If that little function seemes okay to you, my next question is how to implement trim_timepoints into my pytorch-lightning data module? Would something along the lines of the following work?
```python
class fmriDataModule(pl.LightningDataModule):
    def __init__(self, batch_size, csv_file, data_dir):
        super().__init__()
        self.csv_file = csv_file
        self.data_dir = data_dir
        self.batch_size = batch_size
    
    def trim_timepoints(self, image, frames):
        reshape1 = tio.Lambda(lambda x: x.permute(1, 2, 3, 0))
        reshape2 = tio.Lambda(lambda x: x.permute(3, 0, 1, 2))
        img = tio.ScalarImage(image)
        w, d, h, t = img[tio.DATA].size()
        crop = tio.CropOrPad((h, frames, w))
        reshaped = reshape1(img)
        cropped = crop(reshaped)
        new = reshape2(cropped)
        return new

    def get_augmentation_transform(self):
        augment = tio.Compose([
            tio.RandomAffine(p=0.2),
            tio.RandomFlip(p = 0.2),
            tio.RandomGamma(p=0.2),
            tio.RandomNoise(p=0.2),
            tio.RandomMotion(p=0.2),
            tio.RandomBiasField(p=0.2),
        ])
        return augment

    def setup(self, stage=None):
        
        subjList = fmriDataset(csv_file = self.csv_file, root_dir = self.data_dir)
    
        train_size, val_size = int(0.7 * len(subjList)), int(0.2 * len(subjList))
        test_size = len(subjList) - train_size - val_size
        
        if stage == 'fit' or stage is None:
            self.train_dataset, self.val_dataset, _ = torch.utils.data.random_split(subjList, [train_size, val_size, test_size])

        if stage == 'test' or stage is None:
        _, _, self.test_dataset = torch.utils.data.random_split(subjList, [train_size, val_size, test_size])

        augment = self.get_augmentation_transform()
        trim_timepoints = self.trim_timepoints()

        train_transforms = tio.Compose([trim_timepoints, augment])

        self.train_set = tio.SubjectsDataset(self.train_dataset, transform=train_transforms)
        self.val_set = tio.SubjectsDataset(self.val_dataset, transform=trim_timepoints)
        self.test_set = tio.SubjectsDataset(self.test_dataset, transform=trim_timepoints)

    def train_dataloader(self):
        return DataLoader(self.train_set, self.batch_size, shuffle=True, num_workers=27)

    def val_dataloader(self):
        return DataLoader(self.val_set, self.batch_size, num_workers=27)

    def test_dataloader(self):
        return DataLoader(self.test_set, self.batch_size, num_workers=27)
```

Main problem is I don't know how define the `image` and `frames` arguments in the trim_timepoints function within the data module.

Thanks for your help!

**Reply by fepegar** [reply]
You could define a `TrimTimepoints` class that inherits from `SpatialTransform`, as e.g., `ToCanonical` does. How do you choose the value for `frames`?

**Reply by ik362** [reply]
Hi @fepegar,

Thanks again for your help/ideas. So far I have come up with the following code:
```python
from torchio.transforms import SpatialTransform
from torchio.data import Subject

class trimTimepoints(SpatialTransform):
    def apply_transform(self, frames, subject: Subject) -> Subject:
        for image in subject.get_images(intensity_only = False):
            reshape1 = tio.Lambda(lambda x: x.permute(1, 2, 3, 0))
            reshape2 = tio.Lambda(lambda x: x.permute(3, 0, 1, 2))
            img = tio.ScalarImage(image)
            w, d, h, t = img[tio.DATA].size()
            crop = tio.CropOrPad((h, frames, w))
            reshaped = reshape1(img)
            cropped = crop(reshaped)
            new = reshape2(cropped)
            image.set_data(torch.tensor(new))
    return subject
```

The code works until the last line where I get:
```python
RuntimeError: Could not infer dtype of ScalarImage
```
How can I fix this error?
<br><br>
> How do you choose the value for frames?

This is specific to my data. It could be automated similar to the tutorial but for my purposes a the moment manually setting this hyperparameter is more suitable.

Thanks!

**Reply by fepegar** [reply]
You're close! `apply_transform` should get only the `subject` argument. You can then get the images using `self.get_images`, which is a method defined in `SpatialTransform`. You can pass the number of frames when you instantiate the transform, as in `CropOrPad` for example. You would store it as an attribute and use it later as `self.frames`.

The error you're getting is because you are passing new, an instance of `ScalarImage`, to `torch.tensor`. You don't need that. You can just do `image.set_data(new.data)`.

**Reply by fepegar** [reply]
Also, I'm not sure why you added `img = tio.ScalarImage(image)`.
